File: tools/tf-types-ts/src/core/format.ts
/**
 * Binary framing for .tflog and .tfproof. Matches
 * `crates/tf-types/src/format.rs` byte-for-byte via
 * `conformance/framing-vectors.yaml`.
 *
 * .tflog  — append-only log of proof events.
 *   header  = "TFLOG\x01\x00\x00"   (8 bytes)
 *   frames  = u32 BE length + canonical-JSON event bytes (repeat)
 *
 * .tfproof — signed bundle container.
 *   header  = "TFPROOF\x01"           (8 bytes)
 *   body    = u32 BE length + canonical-JSON bundle bytes
 *   trailer = u32 BE length + raw signature bytes
 */

import type { ProofBundle } from "../generated/proof-bundle.js";
import type { ProofEvent } from "../generated/proof-event.js";
import { canonicalize } from "./canonical.js";
import { utf8decode, utf8encode } from "./crypto.js";

export const TFLOG_MAGIC = new Uint8Array([0x54, 0x46, 0x4c, 0x4f, 0x47, 0x01, 0x00, 0x00]);
export const TFPROOF_MAGIC = new Uint8Array([0x54, 0x46, 0x50, 0x52, 0x4f, 0x4f, 0x46, 0x01]);

export class FormatError extends Error {}

function putU32BE(view: number[], n: number): void {
  if (n < 0 || n > 0xffffffff) throw new FormatError(`length out of range: ${n}`);
  view.push((n >>> 24) & 0xff, (n >>> 16) & 0xff, (n >>> 8) & 0xff, n & 0xff);
}

function readU32BE(buf: Uint8Array, off: number): number {
  if (off + 4 > buf.length) throw new FormatError(`truncated at offset ${off}`);
  return ((buf[off]! << 24) | (buf[off + 1]! << 16) | (buf[off + 2]! << 8) | buf[off + 3]!) >>> 0;
}

function toBuf(parts: number[] | Uint8Array[]): Uint8Array {
  let total = 0;
  for (const p of parts) total += (p as any).length;
  const out = new Uint8Array(total);
  let off = 0;
  for (const p of parts) {
    if (Array.isArray(p)) {
      for (const b of p) {
        out[off++] = b;
      }
    } else {
      out.set(p as Uint8Array, off);
      off += (p as Uint8Array).length;
    }
  }
  return out;
}

// ---------- .tflog ----------

export function writeTflog(events: readonly ProofEvent[]): Uint8Array {
  const chunks: Uint8Array[] = [TFLOG_MAGIC];
  for (const e of events) {
    const body = utf8encode(canonicalize(e));
    const len: number[] = [];
    putU32BE(len, body.length);
    chunks.push(new Uint8Array(len));
    chunks.push(body);
  }
  return toBuf(chunks);
}

export function appendTflog(existing: Uint8Array, event: ProofEvent): Uint8Array {
  if (existing.length === 0) {
    return writeTflog([event]);
  }
  for (let i = 0; i < TFLOG_MAGIC.length; i++) {
    if (existing[i] !== TFLOG_MAGIC[i]) throw new FormatError("bad .tflog magic");
  }
  const body = utf8encode(canonicalize(event));
  const out = new Uint8Array(existing.length + 4 + body.length);
  out.set(existing, 0);
  const view = new DataView(out.buffer, out.byteOffset + existing.length, 4);
  view.setUint32(0, body.length, false);
  out.set(body, existing.length + 4);
  return out;
}

export function readTflog(buf: Uint8Array): ProofEvent[] {
  if (buf.length < TFLOG_MAGIC.length) throw new FormatError("truncated magic");
  for (let i = 0; i < TFLOG_MAGIC.length; i++) {
    if (buf[i] !== TFLOG_MAGIC[i]) throw new FormatError("bad .tflog magic");
  }
  const out: ProofEvent[] = [];
  let off = TFLOG_MAGIC.length;
  while (off < buf.length) {
    const len = readU32BE(buf, off);
    off += 4;
    if (off + len > buf.length) throw new FormatError(`truncated frame at offset ${off - 4}`);
    const text = utf8decode(buf.subarray(off, off + len));
    out.push(JSON.parse(text) as ProofEvent);
    off += len;
  }
  return out;
}

// ---------- .tfproof ----------

export interface TfproofParts {
  readonly bundle: ProofBundle;
  readonly signature: Uint8Array;
  readonly canonicalBody: Uint8Array;
}

export function writeTfproof(bundle: ProofBundle, signature: Uint8Array): Uint8Array {
  const body = utf8encode(canonicalize(bundle));
  const len1: number[] = [];
  putU32BE(len1, body.length);
  const len2: number[] = [];
  putU32BE(len2, signature.length);
  return toBuf([TFPROOF_MAGIC, new Uint8Array(len1), body, new Uint8Array(len2), signature]);
}

export function readTfproof(buf: Uint8Array): TfproofParts {
  if (buf.length < TFPROOF_MAGIC.length) throw new FormatError("truncated magic");
  for (let i = 0; i < TFPROOF_MAGIC.length; i++) {
    if (buf[i] !== TFPROOF_MAGIC[i]) throw new FormatError("bad .tfproof magic");
  }
  let off = TFPROOF_MAGIC.length;
  const bodyLen = readU32BE(buf, off);
  off += 4;
  if (off + bodyLen > buf.length) throw new FormatError("truncated body");
  const canonicalBody = buf.subarray(off, off + bodyLen);
  const bundle = JSON.parse(utf8decode(canonicalBody)) as ProofBundle;
  off += bodyLen;
  const sigLen = readU32BE(buf, off);
  off += 4;
  if (off + sigLen > buf.length) throw new FormatError("truncated signature");
  const signature = buf.slice(off, off + sigLen);
  return { bundle, signature, canonicalBody };
}

File: crates/tf-types/src/format.rs
//! Binary framing for `.tflog` and `.tfproof`. Matches
//! `tools/tf-types-ts/src/core/format.ts` byte-for-byte via
//! `conformance/framing-vectors.yaml`.
//!
//! `.tflog`  — append-only log of proof events.
//!   header  = "TFLOG\x01\x00\x00"   (8 bytes)
//!   frames  = u32 BE length + canonical-JSON event bytes (repeat)
//!
//! `.tfproof` — signed bundle container.
//!   header  = "TFPROOF\x01"           (8 bytes)
//!   body    = u32 BE length + canonical-JSON bundle bytes
//!   trailer = u32 BE length + raw signature bytes

use serde_json::Value;

use crate::canonical::canonicalize;
use crate::generated::proof_bundle::ProofBundle;
use crate::generated::proof_event::ProofEvent;

pub const TFLOG_MAGIC: &[u8; 8] = b"TFLOG\x01\x00\x00";
pub const TFPROOF_MAGIC: &[u8; 8] = b"TFPROOF\x01";

#[derive(Debug, thiserror::Error, PartialEq, Eq)]
pub enum FormatError {
    #[error("unexpected end of input at offset {0}")]
    Truncated(usize),
    #[error("bad magic header at offset {0}")]
    BadMagic(usize),
    #[error("length prefix {1} exceeds remaining bytes at offset {0}")]
    BadLength(usize, u32),
    #[error("canonical JSON error: {0}")]
    Canonical(String),
    #[error("serde error: {0}")]
    Serde(String),
    #[error("utf-8 error: {0}")]
    Utf8(String),
}

fn put_u32_be(buf: &mut Vec<u8>, n: usize) -> Result<(), FormatError> {
    let n = u32::try_from(n).map_err(|_| FormatError::BadLength(buf.len(), u32::MAX))?;
    buf.extend_from_slice(&n.to_be_bytes());
    Ok(())
}

fn read_u32_be(buf: &[u8], off: usize) -> Result<u32, FormatError> {
    if off + 4 > buf.len() {
        return Err(FormatError::Truncated(off));
    }
    let arr: [u8; 4] = buf[off..off + 4].try_into().unwrap();
    Ok(u32::from_be_bytes(arr))
}

// ---------- .tflog ----------

pub fn write_tflog(events: &[ProofEvent]) -> Result<Vec<u8>, FormatError> {
    let mut buf = Vec::with_capacity(256);
    buf.extend_from_slice(TFLOG_MAGIC);
    for e in events {
        let json = serde_json::to_value(e).map_err(|e| FormatError::Serde(e.to_string()))?;
        let body = canonicalize(&json).map_err(|e| FormatError::Canonical(e.to_string()))?;
        let bytes = body.into_bytes();
        put_u32_be(&mut buf, bytes.len())?;
        buf.extend_from_slice(&bytes);
    }
    Ok(buf)
}

pub fn append_tflog(existing: &mut Vec<u8>, event: &ProofEvent) -> Result<(), FormatError> {
    if existing.is_empty() {
        existing.extend_from_slice(TFLOG_MAGIC);
    } else if existing.len() < TFLOG_MAGIC.len() || &existing[..TFLOG_MAGIC.len()] != TFLOG_MAGIC {
        return Err(FormatError::BadMagic(0));
    }
    let json = serde_json::to_value(event).map_err(|e| FormatError::Serde(e.to_string()))?;
    let body = canonicalize(&json).map_err(|e| FormatError::Canonical(e.to_string()))?;
    let bytes = body.into_bytes();
    put_u32_be(existing, bytes.len())?;
    existing.extend_from_slice(&bytes);
    Ok(())
}

pub fn read_tflog(buf: &[u8]) -> Result<Vec<ProofEvent>, FormatError> {
    if buf.len() < TFLOG_MAGIC.len() {
        return Err(FormatError::Truncated(0));
    }
    if &buf[..TFLOG_MAGIC.len()] != TFLOG_MAGIC {
        return Err(FormatError::BadMagic(0));
    }
    let mut out = Vec::new();
    let mut off = TFLOG_MAGIC.len();
    while off < buf.len() {
        let len = read_u32_be(buf, off)? as usize;
        off += 4;
        if off + len > buf.len() {
            return Err(FormatError::BadLength(off - 4, len as u32));
        }
        let slice = &buf[off..off + len];
        let text = std::str::from_utf8(slice).map_err(|e| FormatError::Utf8(e.to_string()))?;
        let value: Value = serde_json::from_str(text).map_err(|e| FormatError::Serde(e.to_string()))?;
        let event: ProofEvent =
            serde_json::from_value(value).map_err(|e| FormatError::Serde(e.to_string()))?;
        out.push(event);
        off += len;
    }
    Ok(out)
}

// ---------- .tfproof ----------

pub fn write_tfproof(bundle: &ProofBundle, signature: &[u8]) -> Result<Vec<u8>, FormatError> {
    let mut buf = Vec::with_capacity(1024);
    buf.extend_from_slice(TFPROOF_MAGIC);
    let body_json = serde_json::to_value(bundle).map_err(|e| FormatError::Serde(e.to_string()))?;
    let body = canonicalize(&body_json).map_err(|e| FormatError::Canonical(e.to_string()))?;
    let body_bytes = body.into_bytes();
    put_u32_be(&mut buf, body_bytes.len())?;
    buf.extend_from_slice(&body_bytes);
    put_u32_be(&mut buf, signature.len())?;
    buf.extend_from_slice(signature);
    Ok(buf)
}

#[derive(Debug)]
pub struct TfproofParts {
    pub bundle: ProofBundle,
    pub signature: Vec<u8>,
    pub canonical_body: Vec<u8>,
}

pub fn read_tfproof(buf: &[u8]) -> Result<TfproofParts, FormatError> {
    if buf.len() < TFPROOF_MAGIC.len() {
        return Err(FormatError::Truncated(0));
    }
    if &buf[..TFPROOF_MAGIC.len()] != TFPROOF_MAGIC {
        return Err(FormatError::BadMagic(0));
    }
    let mut off = TFPROOF_MAGIC.len();
    let body_len = read_u32_be(buf, off)? as usize;
    off += 4;
    if off + body_len > buf.len() {
        return Err(FormatError::BadLength(off - 4, body_len as u32));
    }
    let body_slice = &buf[off..off + body_len];
    let body_text =
        std::str::from_utf8(body_slice).map_err(|e| FormatError::Utf8(e.to_string()))?;
    let body_value: Value =
        serde_json::from_str(body_text).map_err(|e| FormatError::Serde(e.to_string()))?;
    let bundle: ProofBundle =
        serde_json::from_value(body_value).map_err(|e| FormatError::Serde(e.to_string()))?;
    off += body_len;

    let sig_len = read_u32_be(buf, off)? as usize;
    off += 4;
    if off + sig_len > buf.len() {
        return Err(FormatError::BadLength(off - 4, sig_len as u32));
    }
    let signature = buf[off..off + sig_len].to_vec();

    Ok(TfproofParts {
        bundle,
        signature,
        canonical_body: body_slice.to_vec(),
    })
}
